cmake/CMakeLists.txt

```
@@ -156,15 +165,15 @@ endif()
 ######################################################
 # packages with special compiler needs or external libs
 ######################################################
-include_directories(${LAMMPS_SOURCE_DIR})
+target_include_directories(lammps PRIVATE ${LAMMPS_SOURCE_DIR})
```

Would it be possible to add these to the build-interface so one can embed lammps via `add_subdirectory`?
```suggestion
target_include_directories(lammps PUBLIC $<BUILD_INTERFACE:${LAMMPS_SOURCE_DIR}>)
```
Should probably be added to all `target_include_directories` expressions, so far i've tested it with `${LAMMPS_SOURCE_DIR}/stubs`, `${${PKG}_SOURCES_DIR` and `${LAMMPS_STYLE_HEADERS_DIR}`.

---

> Would it be possible to add these to the build-interface so one can embed lammps via `add_subdirectory`?
> 
> Should probably be added to all `target_include_directories` expressions, so far i've tested it with `${LAMMPS_SOURCE_DIR}/stubs`, `${${PKG}_SOURCES_DIR` and `${LAMMPS_STYLE_HEADERS_DIR}`.

Maybe in a future PR.

---

Setting them `PRIVATE` explicitely prohibits all forwarding, doesn't it?
If i understand it correctly, `$<BUILD_INTERFACE:...>` should only expand when lammps is embedded into the build-process (e.g. via `add_subdirectory`), not if it's installed and included via `find_package`, so it's not necessarily part of your public API.

---

Actually I had another testing with `examples/COUPLE/simple` and you are right be need at least this one, I am not so sure about all the package libs.

---

> Setting them `PRIVATE` explicitely prohibits all forwarding, doesn't it?
> If i understand it correctly, `$<BUILD_INTERFACE:...>` should only expand when lammps is embedded into the build-process (e.g. via `add_subdirectory`), not if it's installed and included via `find_package`, so it's not necessarily part of your public API.

The other problem is I don't want every include path everywhere hence I made them all `PRIVATE` for now. I think the better solution is if you explicitly using e.g. `atc` just add `target_link_libraries(<your_target> atc)` to your code. The `lammps` target should only expose what is explicitly covered by the API.

---

I definitely support your intent, though i see a difference between using a installed version of lammps and embedding it. In the latter case, one has complete control anyways, so it's mostly annoying to be prohibited from using internals. The furthest I'd go is adding a separate target (e.g. `lammps_internals`) that makes clear the api is not stable/support beyond this point.
That said, it's mostly discoverability i'm missing. Is there any way that one can query installed styles without including the `style_*.h` headers?

---

> That said, it's mostly discoverability i'm missing. Is there any way that one can query installed styles without including the `style_*.h` headers?

We currently have `lammps_config_has_package`. We could add a `lammps_config_has_style(type, name)`, and/or `lammps_config_style_count(type)`, `lammps_config_style_name(type, idx)`, similar to the `lammps_config_package_count` and `lammps_config_package_name()` functions.

---

@sgsaenger the question is: do you need to discover the installed packages/styles at runtime or at compile time and do you use the C++ API or the C (library) API?

---

That would be great!

---

@akohlmey I'm using the C++ API. I don't care whether it's compile or runtime, i'm happy either ways. Maybe an array such as `LAMMPS::installed_packages` could be created and serve as a backend for corresponding `library.h` functions?

---

@sgsaenger that is literally what `lammps_config_package_count` and `lammps_config_package_name` are for.

---

Yes, what i meant: there could be a `LAMMPS::installed_styles` that backs up your proposed `lammps_config_style_name`, sorry for the confusion.

---

Ok, well all the infrastructure is already there (it's multiple maps), we only need to expose it in a consistent way. I'll talk this through with the rest of the core devs.

---

@sgsaenger please look at the Info class, e.g. at:
```
// the is_available() function returns true if the selected style
// or name in the selected category is available for use (but need
// not be currently active).

bool Info::is_available(const char *category, const char *name)
```
That and similar calls should already have the info you are looking for. What @rbberger is suggesting would be to expose those APIs to the C library interface.

---

@akohlmey Unfortunately, the info class _itself_ doesn't help with discoverability at all. `Info::is_available` only makes sense when one is looking for a specific style, the `Info::*_styles` functions are private and only print to a file. But it pointed me in the right direction, what i'm actually after are the maps in the specific subclasses, e.g. `Force::pair_map`. Thanks for your help. Maybe it's still possible to expose these maps in a more visible or better documented way?

---

@sgsaenger please let us know what is the use case for what you are looking for. That might help to tailor the interfaces. @rbberger is talking about library functions and APIs he is planning to implement (soon).

---

Out of scope for this PR though.